File: framework/ui/src/io/intino/alexandria/ui/displays/components/Step.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.*;
import io.intino.alexandria.ui.displays.Display;
import io.intino.alexandria.ui.displays.components.stepper.StepChecker;
import io.intino.alexandria.ui.displays.notifiers.StepNotifier;

public class Step<DN extends StepNotifier, B extends Box> extends AbstractStep<B> {
    private StepChecker nextChecker = null;
    private StepChecker backChecker = null;
    private String icon;
    private boolean isActive = false;
    private boolean isCompleted = false;
    private boolean isDisabled = false;

    public Step(B box) {
        super(box);
    }

    public void icon(String icon) {
        _icon(icon);
        refresh();
    }

    protected Step _icon(String icon) {
        this.icon = icon;
        return this;
    }

    protected Step _isActive(boolean isActive) {
        this.isActive = isActive;
        return this;
    }

    protected Step _isDisabled(boolean isDisabled) {
        this.isDisabled = isDisabled;
        return this;
    }

    protected Step _isCompleted(boolean isCompleted) {
        this.isCompleted = isCompleted;
        return this;
    }

    public Step canNext(StepChecker checker){
        this.nextChecker = checker;
        return this;
    }

    public Step canBack(StepChecker checker){
        this.backChecker = checker;
        return this;
    }

    @Override
    public void init() {
        super.init();
        refresh();
    }

    public void contentRendered() {
        children().forEach(Display::refresh);
    }

    @Override
    public void refresh() {
        super.refresh();
        notifier.refresh(
            new StepInfo()
                .icon(this.icon)
                .isActive(this.isActive)
                .isCompleted(this.isCompleted)
                .isDisabled(this.isDisabled)
        );
    }

    protected boolean allowNext(){
        return this.nextChecker == null || this.nextChecker.check(this);
    }

    protected boolean allowBack() {
        return this.backChecker == null || this.backChecker.check(this);
    }
}

File: framework/ui-elements/src/displays/components/Step.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractStep from "../../../gen/displays/components/AbstractStep";
import StepNotifier from "../../../gen/displays/notifiers/StepNotifier";
import StepRequester from "../../../gen/displays/requesters/StepRequester";
import DisplayFactory from 'alexandria-ui-elements/src/displays/DisplayFactory';
import MuiStep from "@material-ui/core/Step";
import MuiStepLabel from "@material-ui/core/StepLabel";
import StepIcon from "./icon/StepIcon";
import { withSnackbar } from 'notistack';
import Delayer from "../../util/Delayer";

const styles = theme => ({});

class Step extends AbstractStep {

	constructor(props) {
		super(props);
		this.notifier = new StepNotifier(this);
		this.requester = new StepRequester(this);
		this.state = {
		    ...this.state,
            icon: undefined,
            isActive: false,
            isDisabled: false,
            isCompleted: false
		};
	};

	render() {
		if (!this.state.visible) return (<React.Fragment/>);
		return (this.props.showContent) ? this._renderContent() : this._renderStep();
	};

	_renderContent(){
		return this.props.children;
	}

	_renderStep(){
		const { index, label } = this.props;
		const { isActive, isDisabled, isCompleted, icon } = this.state;
		const stepProps = { active: isActive, disabled: isDisabled, completed: isCompleted };
		const labelProps = { icon : (icon || index + 1), StepIconComponent: (icon ? StepIcon : undefined) };
		return (
			<MuiStep key={index} {...stepProps}>
				<MuiStepLabel {...labelProps}>{label}</MuiStepLabel>
			</MuiStep>
		);
	};

	componentDidUpdate() {
		if (this.props.showContent) Delayer.execute(this, () => this.requester.contentRendered());
	};

	refresh = (newState) => {
		this.setState(newState);
	};
}

export default withStyles(styles, { withTheme: true })(withSnackbar(Step));
DisplayFactory.register("Step", withStyles(styles, { withTheme: true })(withSnackbar(Step)));